### PR TITLE
Code Tidy November 2019

### PR DESCRIPTION
**Pull request by trisyoungs**
THis PR attempts to win the award for the most boring piece of work so far, and deals with:

- Removing all warnings associated with unused variables.
- Removing many warnings associated to comparison between signed and unsigned ints.
- Replaces some magic numbers with `std::optional` where appropriate.
- Fixes some initialisation order warnings, and moves to initialiser lists where appropriate.
- Removes a couple of other errors related to polymorphic class destructors and potential variable use when uninitialised.
- Sprinkles many `auto` in `for` loops.

Not all signed/unsigned int warnings are removed, since some required more fundamental changes to classes beyond the scope of this PR (e.g. `Module::nTargetConfigurations()`).

**Builds on #412 - do not merge until #412 is resolved.**

### CODE REVIEW

**Review by DanNixon — CHANGES_REQUESTED** [review]
I don't like all the `unsigned`s being thrown around here. See https://google.github.io/styleguide/cppguide.html#Integer_Types for details.

I short, integers should be signed, when iterating using indices the appropriate size type should be used (e.g. std::vector::size_type, or via decltype()).

**Comment by trisyoungs** [comment]
Should have read that first! Reverted the `unsigned int` stuff.

File: src/math/interpolator.cpp [hunk 1]
```diff
@@ -4,7 +4,7 @@
 #include "math/interpolator.h"
 #include "math/data1d.h"
 
-Interpolator::Interpolator(const Array<double> &x, const Array<double> &y, InterpolationScheme scheme) : x_(x), y_(x)
+Interpolator::Interpolator(const Array<double> &x, const Array<double> &y, InterpolationScheme scheme) : x_(x), y_(y)
```
Comment (rprospero) [inline src/math/interpolator.cpp:7]: Good catch

File: src/math/interpolator.cpp [hunk 2]
```diff
@@ -134,14 +134,14 @@ void Interpolator::interpolateSpline()
      * 					x(i) - x(i-1)
      */
 
-    int i, nPoints = x_.nItems();
+    const int nPoints = x_.nItems();
```
Comment (rprospero) [inline src/math/interpolator.cpp:137]: Could this be `const auto`?

File: src/modules/rdf/functions.cpp [hunk 3]
```diff
@@ -486,9 +479,9 @@ bool RDFModule::calculateUnweightedGR(ProcessPool &procPool, Configuration *cfg,
         // 			// Add contributions from this SpeciesIntra only
         // 			std::shared_ptr<const Molecule>* molecules = cfg->molecules().array();
         // 			std::shared_ptr<const Molecule> mol;
-        // 			for (int n=0; n<molecules.nItems(); ++n, mol = molecules.at(n))
+        // 			for (auto n=0; n<molecules.nItems(); ++n, mol = molecules.at(n))
         // 			{
-        // 				for (int n=bondPointers.nItems()-1; n>=0; --n)
+        // 				for (auto n=bondPointers.nItems()-1; n>=0; --n)
         // 				{
```
Comment (rprospero) [inline src/modules/rdf/functions.cpp:485]: While we're cleaning up, do we want to take out all this commented code?

File: src/modules/refine/functions.cpp [hunk 4]
```diff
@@ -70,8 +70,8 @@ bool RefineModule::modifyBondTerms(CoreData &coreData, const Data1D &deltaGR, st
     // exp(-(((x-r)-delta)**2)/width**2)-exp(-(((x-r)+delta)**2)/width**2) w l
 
     // Scan through master bond terms searching for those that involve the AtomTypes specified
-    const auto idI = typeI->index();
-    const auto idJ = typeJ->index();
+    //     const auto idI = typeI->index();
+    //     const auto idJ = typeJ->index();
```
Comment (rprospero) [inline src/modules/refine/functions.cpp:74]: Just delete these lines?

File: src/modules/sq/process.cpp [hunk 5]
```diff
@@ -146,9 +146,9 @@ bool SQModule::process(Dissolve &dissolve, ProcessPool &procPool)
         //             return false;
         //
         //         // Apply necessary broadening
-        //         for (int i = 0; i < unweightedsq.nAtomTypes(); ++i)
+        //         for (auto i = 0; i < unweightedsq.nAtomTypes(); ++i)
         //         {
-        //             for (int j = i; j < unweightedsq.nAtomTypes(); ++j)
+        //             for (auto j = i; j < unweightedsq.nAtomTypes(); ++j)
```
Comment (rprospero) [inline src/modules/sq/process.cpp:151]: Again, should we cut these comments?

**Review by rprospero — APPROVED** [review]
Some pedantic comments, but it's a supposed to be a tidy PR

File: src/math/interpolator.cpp (hunk 1 above)
Comment (trisyoungs) [reply src/math/interpolator.cpp:7]: Thanks. Shame it got there in the first place though!

File: src/math/interpolator.cpp (hunk 2 above)
Comment (trisyoungs) [reply src/math/interpolator.cpp:137]: Yep - got reverted along with the `unsigned int`.

File: src/modules/rdf/functions.cpp (hunk 3 above)
Comment (trisyoungs) [reply src/modules/rdf/functions.cpp:485]: I have been meaning to add an issue to remove all of the stuff related to this, which is an attempt to broaden intramolecular terms based on their fundamental frequency.  It was horrible in the first place, and so I don't mind losing it! I'll add an issue to that effect.

File: src/modules/refine/functions.cpp (hunk 4 above)
Comment (trisyoungs) [reply src/modules/refine/functions.cpp:74]: Similar argument to above, but this time the whole module needs to go!

File: src/modules/sq/process.cpp (hunk 5 above)
Comment (trisyoungs) [reply src/modules/sq/process.cpp:151]: This I would prefer to leave in for the time being, as the very next major bit of work will immediately reinstate this section.

File: src/keywords/nodevalueenumoptions.h [hunk 6]
```diff
@@ -59,9 +59,9 @@ class NodeValueEnumOptionsKeyword : public NodeValueEnumOptionsBaseKeyword, publ
 {
     public:
     NodeValueEnumOptionsKeyword(ProcedureNode *parentNode, NodeValue value, EnumOptions<E> enumOptions)
-        : KeywordData<Venum<NodeValue, E>>(KeywordBase::NodeValueEnumOptionsData, Venum<NodeValue, E>(value, enumOptions)),
-          NodeValueEnumOptionsBaseKeyword(KeywordData<Venum<NodeValue, E>>::data_.value(),
-                                          KeywordData<Venum<NodeValue, E>>::data_.baseOptions())
+        : NodeValueEnumOptionsBaseKeyword(KeywordData<Venum<NodeValue, E>>::data_.value(),
+                                          KeywordData<Venum<NodeValue, E>>::data_.baseOptions()),
+          KeywordData<Venum<NodeValue, E>>(KeywordBase::NodeValueEnumOptionsData, Venum<NodeValue, E>(value, enumOptions))
```
[the remark this answers is not in the document]
Comment (trisyoungs) [reply src/keywords/nodevalueenumoptions.h:64]: Just changed the order to match that of the inheritance heirarchy - even `gcc` warned about this one...

File: src/expression/functions.cpp [hunk 7]
```diff
@@ -93,7 +93,7 @@ ExpressionFunctionData ExpressionFunctions::data[ExpressionFunctions::nFunctions
 // Return enumerated command from string
 ExpressionFunctions::Function ExpressionFunctions::function(std::string_view s)
 {
-    for (int result = ExpressionFunctions::NoFunction; result < ExpressionFunctions::nFunctions; result++)
+    for (auto result = (int)ExpressionFunctions::NoFunction; result < ExpressionFunctions::nFunctions; ++result)
```
[the remark this answers is not in the document]
Comment (trisyoungs) [reply src/expression/functions.cpp:96]: Yes, sounds sensible. Do you want to go ahead an add an issue?

**Review by DanNixon — APPROVED** [review]

**Pull request merged**